homeassistant/components/canary/__init__.py

```diff
@@ -34,33 +41,102 @@
     extra=vol.ALLOW_EXTRA,
 )
 
-CANARY_COMPONENTS = ["alarm_control_panel", "camera", "sensor"]
+PLATFORMS = ["alarm_control_panel", "camera", "sensor"]
+
+
+async def async_setup(hass: HomeAssistantType, config: dict) -> bool:
+    """Set up the Canary integration."""
+    hass.data.setdefault(DOMAIN, {})
+
+    if hass.config_entries.async_entries(DOMAIN):
+        return True
 
+    ffmpeg_arguments = DEFAULT_FFMPEG_ARGUMENTS
+    if CAMERA_DOMAIN in config:
+        camera_config = next(
+            (item for item in config[CAMERA_DOMAIN] if item["platform"] == DOMAIN),
+            None,
+        )
+
+        if camera_config:
+            ffmpeg_arguments = camera_config.get(
+                CONF_FFMPEG_ARGUMENTS, DEFAULT_FFMPEG_ARGUMENTS
+            )
+
+    if DOMAIN in config:
+        if ffmpeg_arguments != DEFAULT_FFMPEG_ARGUMENTS:
+            config[DOMAIN][CONF_FFMPEG_ARGUMENTS] = ffmpeg_arguments
+
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": SOURCE_IMPORT},
+                data=config[DOMAIN],
+            )
+        )
+    return True
 
-def setup(hass, config):
-    """Set up the Canary component."""
-    conf = config[DOMAIN]
-    username = conf[CONF_USERNAME]
-    password = conf[CONF_PASSWORD]
-    timeout = conf[CONF_TIMEOUT]
+
+async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry) -> bool:
+    """Set up Canary from a config entry."""
+    if not entry.options:
+        options = {
+            CONF_FFMPEG_ARGUMENTS: entry.data.get(
+                CONF_FFMPEG_ARGUMENTS, DEFAULT_FFMPEG_ARGUMENTS
+            ),
+        }
+        hass.config_entries.async_update_entry(entry, options=options)
 
     try:
-        hass.data[DATA_CANARY] = CanaryData(username, password, timeout)
-    except (ConnectTimeout, HTTPError) as ex:
-        _LOGGER.error("Unable to connect to Canary service: %s", str(ex))
-        hass.components.persistent_notification.create(
-            f"Error: {ex}<br />You will need to restart hass after fixing.",
-            title=NOTIFICATION_TITLE,
-            notification_id=NOTIFICATION_ID,
+        canary_data = CanaryData(
+            entry.data[CONF_USERNAME],
+            entry.data[CONF_PASSWORD],
+            entry.data[CONF_TIMEOUT],
         )
+    except ConnectTimeout as error:
+        raise ConfigEntryNotReady from error
+    except HTTPError as error:
+        _LOGGER.error("Unable to connect to Canary service: %s", str(error))
```

Comment:
Why would this not be `ConfigEntryNotReady` as well? If the service or upstream proxy isn't working, then trying again later would be good and `ConfigEntryNotReady` will handle the retries and cool down for you.
Doing so will avoid the use having to re-add your integration or restart HASS.

Reply:
ideally we need to detect timeouts vs auth issues but thats a bit trickier with requests lib